runtime/src/bank.rs

```diff
@@ -1809,6 +1832,53 @@ impl Bank {
             });
     }
 
+    /// Get any cached executors needed by the transaction
+    fn get_executors(
+        &self,
+        message: &Message,
+        loaders: &[Vec<(Pubkey, Account)>],
+    ) -> Rc<RefCell<Executors>> {
+        let mut num_executors = message.account_keys.len();
+        for instruction_loaders in loaders.iter() {
+            num_executors += instruction_loaders.len();
+        }
+        let mut executors: HashMap<Pubkey, Arc<dyn Executor>> =
+            HashMap::with_capacity(num_executors);
+        let cached_executors = self.cached_executors.read().unwrap();
+
+        for key in message.account_keys.iter() {
+            if let Some(cached_executor) = cached_executors.get(key) {
+                executors.insert(*key, cached_executor.executor.clone());
+            }
+        }
+        for instruction_loaders in loaders.iter() {
+            for (key, _) in instruction_loaders.iter() {
+                if let Some(cached_executor) = cached_executors.get(key) {
+                    executors.insert(*key, cached_executor.executor.clone());
+                }
+            }
+        }
+
+        Rc::new(RefCell::new(Executors {
+            executors,
+            is_dirty: false,
+        }))
+    }
+
+    /// Add executors back to the bank's cache if modified
+    fn update_executors(&self, executors: Rc<RefCell<Executors>>) {
+        let executors = executors.borrow();
+        if executors.is_dirty {
+            let mut cached_executors = self.cached_executors.write().unwrap();
+            for (key, executor) in executors.executors.iter() {
```

Comment:
`cached_executors` probably should be a LRU cache.  Seems like I could run you out of memory by sending cheap transactions at every single program that has ever existed, to cause them all to get added into `cached_executors`

Reply:
@mvines Or maybe an LFU cache would be best, the problem with the LRU/LFU is that updating the "used" info requires a write lock on `get`.  I switched to a fixed size LRU cache but only `peek` during `get` which does not require a write lock.  This effectively turns the LRU into FIFO cache.  This means we only take a write lock if the program is not in the cache and only for a small amount of time since we are inserting references.  Thoughts?

Reply:
Atomic u64 for timestamp last used, don’t need a lock unless you update the set.

Reply:
@sakridge Looks like most of these implementations re-order a doubly linked list so that LFU ejection doesn't involve a search, though maybe a search during `Finalize` is acceptable

Reply:
Maybe can put together our own LFU using https://docs.rs/plru/0.1.1/plru/index.html

Reply:
Yea.. the search is probably not that bad. Anything that keeps the set sorted seems like you need to take a write lock to update the usage.

Reply:
@sakridge I added a LFU cache for the executors, only requires write lock on put :-)